File: ejTasks/start_here/SimpleServer/SimpleServer.java
package SimpleServer;

import java.io.BufferedReader;
import java.io.Closeable;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.io.OutputStreamWriter;
import java.io.PrintWriter;
import java.net.ServerSocket;
import java.net.Socket;
import java.nio.charset.StandardCharsets;
import java.util.Optional;
import java.util.logging.Level;
import java.util.logging.Logger;

public class SimpleServer implements Closeable {
	private int m_port;
	private Logger m_logger;
	private ServerSocket m_serverSocket;
	private Optional<Connection> m_activeRequest = Optional.empty(); 

	public static SimpleServer create(final int port, final Logger logger) throws IOException {
		return new SimpleServer(port, logger);
	}

	public class Connection implements Closeable {
		private final Socket m_curSocket;
		private final BufferedReader m_reader;
		private final PrintWriter m_writer;

		public Connection(final Socket s) throws IOException {
			m_curSocket = s;

			InputStream input = m_curSocket.getInputStream();
            m_reader = new BufferedReader(new InputStreamReader(input));

			OutputStream output = m_curSocket.getOutputStream();
			m_writer = new PrintWriter(new OutputStreamWriter(output, StandardCharsets.UTF_8), true);
		}

		public String getInput() {
			try {
	            String line = m_reader.readLine();
	            if (line == null) {
	    			m_logger.log(Level.WARNING, "Receive terminated");
	    			close();
	    			return "";
	            }
	            
	            return line;
			}
			catch (IOException ex) {
				m_logger.warning(String.format("Socket error: ", ex.toString()));
				close();
				return "";
			}
		}

		public void write(String a_text) {
			m_logger.log(Level.FINEST, "Server is writing " + a_text);
			m_writer.print(a_text);
			m_writer.flush();
		}

		public void writeln(String a_text) {
			m_logger.log(Level.FINEST, "Server is writing " + a_text);
			m_writer.println(a_text);
			m_writer.flush();
		}

		@Override
		public void close() {
			try {
				m_reader.close();
				m_writer.close();
				m_curSocket.close();
			}
			catch (IOException ex) {
				m_logger.severe(String.format("Error closing socket: ", ex));
			}

			if (m_activeRequest.isPresent() && (m_activeRequest.get() == this)) {
				m_activeRequest = Optional.empty();
			}
		}
	}

	public SimpleServer(int port, Logger logger) throws IOException {
		m_port = port;
		m_logger = logger;
		m_serverSocket = new ServerSocket(port);
 	}
	
	public Connection waitForConnection() throws IOException {
		if (m_activeRequest.isPresent()) {
			throw new PreviousConnectionNotClosed();
		}
		
		Optional<Socket> socket = Optional.empty();
		while (true) {
			try {
				m_logger.log(Level.INFO, "Server is listening on port " + m_port);
	
				socket = Optional.of(m_serverSocket.accept());
				m_logger.log(Level.INFO, "New client connected");
				
				if (!socket.isEmpty() && socket.get() != null) {
					try {
						return new Connection(socket.get());
					}
					catch (IOException ex) {
						m_logger.severe(String.format("Socket error during connection init: ", ex.toString()));
					}
				}
			}
			catch (IOException ex) {
				if (socket.isPresent()) {
					socket.get().close();
				}
				m_logger.log(Level.WARNING, "Socket read error " + ex.getMessage());
				throw ex;
			}
		}
	}

	@Override
	public void close() throws IOException {
		m_serverSocket.close();
	}
}

File: sharpTasks/start_here/simple_server/SimpleServer.cs
using System.Net;
using System.Net.Sockets;
using System.Text;
using Microsoft.Extensions.Logging;
using PiiLib;
using Utils;

namespace SimpleServer
{
    public class SimpleServerFactory {
        private readonly ILogger<SimpleServer> m_logger;

        public SimpleServerFactory(ILogger<SimpleServer> a_logger) {
            m_logger = a_logger;
        }

        public SimpleServer Create(int a_port)
        {
            return new SimpleServer(a_port, m_logger);
        }
    }

    public class SimpleServer : IDisposable
    {
        private readonly TcpListener m_serverSocket;
        private readonly ILogger<SimpleServer> m_logger;
        private Optional<Connection> m_activeRequest = Optional.Empty<Connection>();

        public class Connection : IDisposable
        {
            private readonly SimpleServer m_server;
            private readonly TcpClient m_curSocket;
            private readonly StreamReader m_reader;
            private readonly StreamWriter m_writer;

            public Connection(SimpleServer a_server, TcpClient a_client)
            {
                m_server = a_server;
                m_curSocket = a_client;
                NetworkStream stream = m_curSocket.GetStream();
                m_reader = new StreamReader(stream, Encoding.UTF8);
                m_writer = new StreamWriter(stream, Encoding.ASCII) { AutoFlush = true };
            }

            public string GetInput()
            {
                try
                {
                    string? line = m_reader.ReadLine();
                    if (line == null)
                    {
                        m_server.m_logger.LogWarning("Receive terminated");
                        Close();
                        return "";
                    }

                    return line;
                }
                catch (IOException ex)
                {
                    m_server.m_logger.LogWarning("Socket error {ex}", ex.Message);
                    Close();
                    return "";
                }
            }

            public void Write(PiiConcat text)
            {
                m_server.m_logger.LogDebug("Server is writing {text}", text);
                m_writer.Write(text.ExposeUnsecured());
                m_writer.Flush();
            }

            public void WriteLine(PiiConcat text)
            {
                m_server.m_logger.LogDebug("Server is writing {text}", text);
                m_writer.WriteLine(text.ExposeUnsecured());
                m_writer.Flush();
            }

            public void Close()
            {
                m_reader.Close();
                m_writer.Close();
                m_curSocket.Close();

                if (m_server.m_activeRequest.IsPresent() && m_server.m_activeRequest.Get() == this)
                {
                    m_server.m_activeRequest = Optional.Empty<Connection>();
                }
            }

            public void Dispose()
            {
                Close();
            }
        }

        public SimpleServer(int port, ILogger<SimpleServer> logger)
        {
            m_logger = logger;
            m_serverSocket = new TcpListener(IPAddress.Any, port);
            m_serverSocket.Start();
        }

        public Connection WaitForConnection()
        {
            if (m_activeRequest.IsPresent())
            {
                throw new PreviousConnectionNotClosed();
            }

            TcpClient client = m_serverSocket.AcceptTcpClient();
            m_logger.LogInformation("New client connected");

            try
            {
                return new Connection(this, client);
            }
            catch (IOException ex)
            {
                m_logger.LogCritical("Socket error during connection init: {ex}", ex.Message);
                throw;
            }
        }

        public void Dispose()
        {
            m_serverSocket.Stop();
        }
    }

    public class PreviousConnectionNotClosed : Exception
    {
        public PreviousConnectionNotClosed()
        {
        }
    }

}
